### ISSUE TITLE
Still Active?

### ISSUE THREAD

**Issue by prashnts** [axiak/pybloomfiltermmap#71]
I see several open PRs adding Python 3 support and I was wondering if this project is still being maintained. I could volunteer, though, adding Pyk3 support is just a 2-to-3 transpiler away. :grin:

**Comment by axiak** [comment]
It's in maintenance. I can investigate py3 support. I'm not sure what the interaction between python3 transpiled code and Cython is.

**Comment by prashnts** [comment]
First of all, thanks for this library. From what I have gathered by reading the source code, the library is _already_ almost compatible.
Here's a list of incompatible stuff I see:
- Python 3 has unicode string types by default, so `bytes` type is needed wherever string arguments are involved.
- `Setup.py` does not have parentheses in `print` statements.

Number 2 is no biggie, and Number 1 requires a slight change in API calls: Using `bytes` prefix in string declaration.

This is the modified example:

``` python
>>> from pybloomfilter import BloomFilter
>>> fruit = BloomFilter(100000, 0.1, b'/tmp/words.bloom')
>>> fruit.update((b'apple', b'pear', b'orange', b'apple'))
>>> len(fruit)
3
>>> b'mike' in fruit
False
>>> b'apple' in fruit
True
```

I am not familiar with Cython, but I _can_ live with prefixes.

Maybe you can push a fix for `setup.py` with line number `31` and `43` fixed and make a note about `b"string"` in docs?

**Comment by prashnts** [comment]
Also, reading Cython documentation, it says that this fix should suffice.

``` python
from cpython.version cimport PY_MAJOR_VERSION

cdef unicode _ustring(s):
    if type(s) is unicode:
        # fast path for most common case(s)
        return <unicode>s
    elif PY_MAJOR_VERSION < 3 and isinstance(s, bytes):
        # only accept byte strings in Python 2.x, not in Py3
        return (<bytes>s).decode('ascii')
    elif isinstance(s, unicode):
        # an evil cast to <unicode> might work here in some(!) cases,
        # depending on what the further processing does.  to be safe,
        # we can always create a copy instead
        return unicode(s)
    else:
        raise TypeError(...)
```

But, I am not familiar with Cython so I am unsure where exactly would this go.

http://docs.cython.org/src/tutorial/strings.html

**Comment by endrebak** [comment]
Desperately wanting this also. Since this lib is pretty stable you could probably just create a new python 3 version by changing a few lines and adding it to PyPI?

**Comment by endrebak** [comment]
The only thing you need to change is to add parens in the two print statements in setup.py.

That is all that is needed! You can update the docs with the python3 version of the example:

```
from pybloomfilter import BloomFilter

bf = BloomFilter(10000000, 0.01, b'filter.bloom')

with open("/usr/share/dict/words") as f:
    for word in f:
        bf.add(str.encode(word.rstrip()))

print(b'apple' in bf)
print(str.encode("apple") in bf)
#outputs True
```

Git cloned the repo and ran the modified setup.py and it all works.

I am using pybloomfiltermmap to write a package that I'm uploading to PyPI so hope you can add these 4 characters to the setup.py file soon and upload. If not I'll have to add your source to my package :/

**Comment by prashnts** [comment]
@endrebak Hey! I'm currently updating the package so the workaround isn't needed. It's forked at https://github.com/PrashntS/pybloomfiltermmap3
I can upload the version on PyPI if it's okay.

**Comment by prashnts** [comment]
I'd close this issue, though. @axiak please feel free to reopen. I'm not adding a PR, since I am removing Python 2 support in my fork.

**Comment by endrebak** [comment]
@PrashntS Please remember to add an issues page to the fork (I think it is under "settings"). It is not done automatically on forked repos :)

**Comment by prashnts** [comment]
@endrebak I did not know that! Thanks. I've added it. So far, 22 of 24 unittests are passing. I'll open an issue in the fork.

**Comment by axiak** [comment]
OK @endrebak @PrashntS I released 0.3.15 with the suggested fixes for Python3. Let me know if there's anything else blocking py3.

**Comment by prashnts** [comment]
@axiak I can see quite a few things. `base64` is moved to its own module in py3. String package also has substantial API changes. I have pushed a partial fix, but it is failing at the moment.

https://github.com/PrashntS/pybloomfiltermmap3/commit/fe69de349b7a9a1df591fd28426d0de4286f02c1#diff-228783772f98e1efa594230ccff6bcee
